Index: LayoutTests/http/tests/resources/testharness-helpers.js
diff --git a/testing/web-platform/tests/resources/testharness-helpers.js b/testing/web-platform/tests/resources/testharness-helpers.js
index 292e5f9953ec30c36731d045f504806c0b0c17bb..29d33018fdacf3531235f0a6dcd5521a1d95f208 100644
--- a/testing/web-platform/tests/resources/testharness-helpers.js
+++ b/testing/web-platform/tests/resources/testharness-helpers.js
@@ -108,7 +108,7 @@ self.assert_object_equals = function(actual, expected, description) {
   }
 
   _is_equal(actual, expected,
-            (description ? description + ': ' : '') + _brand(actual));
+            (description ? description + ': ' : '') + _brand(expected));
 };
 
 // Equivalent to assert_in_array, but uses a weaker equivalence relation
